lib/Parse/ParseExpr.cpp

```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
   return makeParserResult(new (Context) PlatformVersionConstraintAvailabilitySpec(
       Platform.getValue(), PlatformLoc, Version, VersionRange));
 }
+
+/// expr-pound-assert:
+///   '#assert' '(' expr ',' string_literal ')'
+ParserResult<Expr> Parser::parseExprPoundAssert() {
+  SyntaxParsingContext AssertContext(SyntaxContext,
+      SyntaxKind::PoundAssertExpr);
+
+  SourceLoc startLoc = consumeToken(tok::pound_assert);
+  SourceLoc endLoc;
+
+  auto errorAndSkipToEnd = [&]() -> ParserResult<Expr> {
+    skipUntilDeclStmtRBrace(tok::r_paren);
+    if (Tok.is(tok::r_paren)) {
+      endLoc = consumeToken();
+    } else {
+      endLoc = PreviousLoc;
+    }
+    return makeParserResult<Expr>(new (Context)
+                                      ErrorExpr(SourceRange(startLoc, endLoc)));
+  };
+
+  if (!Context.LangOpts.EnableExperimentalStaticAssert) {
+    diagnose(startLoc, diag::pound_assert_disabled);
+    return errorAndSkipToEnd();
+  }
+
+  if (parseToken(tok::l_paren, diag::pound_assert_expected, "(")) {
+    return errorAndSkipToEnd();
+  }
+
+  auto conditionExprResult = parseExpr(diag::pound_assert_expected_expression);
+  if (conditionExprResult.isParseError()) {
+    return errorAndSkipToEnd();
+  }
+
+  StringRef message = "assertion failed";
```

Comment:
I don't think Parser should have the default message here. Could you make the default `StringRef()` and handle it in the actual assertion error logic?